### PR TITLE
[IIDM v1.4] Add ThreadLocalMultipleVariantContext (#183)

### LINKED ISSUE #183: Support Variants in multi-threading environment
* **Do you want to request a *feature* or report a *bug*?**
feature

* **What is the current behavior?**
Creating and editing variants of a network is not possible in a multi-threading environment.

* **What is the expected behavior?**
We should be able to create a variant of the network for each thread and then let them modify their variant without any conflict with others threads.

See `ThreadLocalMultiVariantContext` in java

### PR DESCRIPTION
**Pull request by sebalaig**
Signed-off-by: Sébastien LAIGRE <slaigre@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
Closes #183 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: src/iidm/ThreadLocalMultipleVariantContext.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return *m_index;
+    }
+    throw PowsyblException(stdcxx::format("Variant index not set for current thread %1%", std::this_thread::get_id()));
+}
+
+bool ThreadLocalMultipleVariantContext::isIndexSet() const {
+    return static_cast<bool>(m_index);
+}
+
+void ThreadLocalMultipleVariantContext::reset() {
```
Comment (sebalaig) [inline src/iidm/ThreadLocalMultipleVariantContext.cpp:32]: This function exists in java but not called. It should probably be removed;

File: src/iidm/VariantManager.cpp [hunk 2]
```diff
@@ -53,6 +55,26 @@ void VariantManager::allocateVariantArrayElement(unsigned long sourceIndex, cons
...
 
+void VariantManager::allowVariantMultiThreadAccess(bool allow) {
+    if (allow && !stdcxx::isInstanceOf<ThreadLocalMultipleVariantContext>(m_variantContext)) {
+        std::unique_ptr<VariantContext> newVariantContext = stdcxx::make_unique<VariantContext, ThreadLocalMultipleVariantContext>();
+        // For multithreaded VariantContext, don't set the variantIndex to a default
+        // value if it is not set, so that missing initializations fail fast.
+        if (m_variantContext->isIndexSet()) {
+            newVariantContext->setVariantIndex(m_variantContext->getVariantIndex());
+        }
+        m_variantContext.reset(newVariantContext.release());
```
Comment (mathbagu) [inline src/iidm/VariantManager.cpp:66]: Use the assignment operator or the swap method instead?

File: test/iidm/CMakeLists.txt [hunk 3]
```diff
@@ -58,7 +58,7 @@ set(UNIT_TEST_SOURCES
 
 add_executable(unit-tests-iidm ${UNIT_TEST_SOURCES})
 target_compile_definitions(unit-tests-iidm PRIVATE BOOST_TEST_DYN_LINK)
-target_link_libraries(unit-tests-iidm PRIVATE iidm-tests)
+target_link_libraries(unit-tests-iidm PRIVATE iidm-tests Threads::Threads)
```
Comment (mathbagu) [inline test/iidm/CMakeLists.txt:61]: I think this should be add to the IIDM library instead and then inherited by the other targets?

File: test/iidm/VariantManagerTest.cpp [hunk 4]
```diff
@@ -5,6 +5,9 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
+#include <iostream>
```
Comment (mathbagu) [inline test/iidm/VariantManagerTest.cpp:8]: Remove this include

**Review by mathbagu — CHANGES_REQUESTED** [review]

**Review by mathbagu — APPROVED** [review]

**Pull request merged**